src/ZkFold/Base/Data/Vector.hs

```diff
@@ -119,71 +113,60 @@ rotate (Vector lst) n = Vector (r <> l)
         lshift :: Int
         lshift = fromIntegral $ n `mod` len
 
-        (l, r) = P.splitAt lshift lst
+        (l, r) = V.splitAt lshift lst
 
 shift :: forall size a. KnownNat size => Vector size a -> Integer -> a -> Vector size a
 shift (Vector lst) n pad
-  | n < 0 = Vector $ ZP.take (value @size) (padList <> lst)
-  | otherwise = Vector $ ZP.drop (fromIntegral n) (lst <> padList)
+  | n < 0 = Vector $ V.take (knownNat @size) (padList <> lst)
+  | otherwise = Vector $ V.drop (fromIntegral n) (lst <> padList)
     where
-        padList = replicate (fromIntegral $ abs n) pad
-
+        padList = V.replicate (fromIntegral $ abs n) pad
 
 vectorDotProduct :: forall size a . Semiring a => Vector size a -> Vector size a -> a
 vectorDotProduct (Vector as) (Vector bs) = sum $ zipWith (*) as bs
 
 empty :: Vector 0 a
-empty = Vector []
+empty = Vector V.empty
 
 infixr 5 .:
 (.:) :: a -> Vector n a -> Vector (n + 1) a
-a .: (Vector lst) = Vector (a : lst)
+a .: (Vector lst) = Vector (a `V.cons` lst)
 
 append :: Vector m a -> Vector n a -> Vector (m + n) a
 append (Vector l) (Vector r) = Vector (l <> r)
 
 concat :: Vector m (Vector n a) -> Vector (m * n) a
-concat = Vector . concatMap fromVector
+concat = Vector . V.concatMap toV . toV
 
 unsafeConcat :: forall m n a . [Vector n a] -> Vector (m * n) a
-unsafeConcat = Vector . concatMap fromVector
+unsafeConcat = concat . unsafeToVector @m
 
 chunks :: forall m n a . KnownNat n => Vector (m * n) a -> Vector m (Vector n a)
-chunks (Vector lists) = Vector (Vector <$> chunksOf (fromIntegral $ value @n) lists)
+chunks (Vector vectors) = unsafeToVector (Vector <$> V.chunksOf (fromIntegral $ value @n) vectors)
 
 instance Binary a => Binary (Vector n a) where
     put = put . fromVector
     get = Vector <$> get
```

Comment:
There is no guarantee that the parsed ByteString would contain exactly `n` elements, something like `replicate n` is needed here

Reply:
fixed it to use V.replicateM